cmd/kube-rbac-proxy/app/kube-rbac-proxy.go

```diff
@@ -232,7 +232,7 @@ func Run(cfg *server.KubeRBACProxyConfig) error {
 
 	handler := identityheaders.WithAuthHeaders(proxy, cfg.KubeRBACProxyInfo.UpstreamHeaders)
 	handler = kubefilters.WithAuthorization(handler, authz, scheme.Codecs)
-	handler = kubefilters.WithAuthentication(handler, authenticator, http.HandlerFunc(filters.UnauthorizedHandler), cfg.DelegatingAuthentication.APIAudiences)
+	handler = kubefilters.WithAuthentication(handler, authenticator, http.HandlerFunc(filters.UnauthorizedHandler), cfg.DelegatingAuthentication.APIAudiences, nil)
```

Comment:
What happens when you supply the request header nil here? It won't get autodetected from the cluster, right?

Reply:
TODO add explanation of what nil does.

Reply:
It enables you to cleanse custom headers that you use for your proxy to contain additional information. It is necessary to drop those.

We write custom headers in case of request headers and we read custom headers based on the rewrite logic: [link](https://github.com/ibihim/kube-rbac-proxy/blob/f31c1870b1800a92f0f37c01b7a20c17adfc5412/pkg/authorization/rewrite/middleware.go#L56-L63).
The rewrite logic for Query parameters relies on upstream to interpret them, which is usually a hostile behavior that you want to avoid, but in our case, we want Prometheus to be able to interpret it. Therefore I am hesitant to introduce it for custom headers as well. We could add a boolean flag that defaults to true, which enables / disables cleansing of header values? **WDYT @stlaz?**

Ref: [k8s.io/apiserver/pkg/endpoints/filters/authentication.go](https://github.com/kubernetes/kubernetes/blob/1904a56e3071035de19016cef621866a37140d5f/staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go#L98-L105)

Reply:
We should avoid forwarding the default headers at least, which is what `WithAuthentication()`does already.

Generally I think we should drop any headers we're processing, more so if we're using them for authentication. I don't think we wired any authn via request headers, correct?

Outside of that we should also make sure that none of the headers we produce can be just passed upstream from the original request.

Reply:
Let me summarize:

- `WithAuthentication` does a great job.
- We should remove headers we consume, which would include the rewrites rule in the HEADER. This task would include some research on the impact on Prometheus.
- If identityheaders are enabled, we should remove those headers if set by the client. This is also a request from the review by Mo.

This would be another PR.

Reply:
At least add a FIXME here then, I think we need to address this before this makes it to production

Reply:
Fixing before production? 100%. I can add a FIXME.